Unify HTTP response model and API

FrankChen021 opened the pull request:
### Motivation

Current HTTP APIs return different formats of response. In most cases, if the status code is 200, a JSON object is returned. If the status code is not 200, there're several formats, some return a JSON object, some return text. For those APIs that return text, there might be vulnerabilities that allows hackers to launch an XSS attack.

In order to make code safe, this PR unifies the API that construct a safe Response object. 

After this PR, for those HTTP responses whose Status Code is not 200:
1. there's alway a Content-Type in the response header which is set to 'application/json'
2. and the HTTP body is a JSON object which currently has one field named as "error".

```json
{"error" : "xxxx"}
```

In order to make sure the new API is always used in the future code, the raw `javax.ws.rs.core.Response#status` API is forbidden.

### API Description

The New API is defined as `HttpResponses`. Here's an example that compares how we construct a Response object in old and new API.

#### Old ways

```java
return Response.status(Response.Status.BAD_REQUEST)
                     .entity(ImmutableMap.<String, Object>of(
                         "error",
                         StringUtils.format(UNKNOWN_AUTHENTICATOR_MSG_FORMAT, authenticatorName)
                     ))
                     .build();
```

#### New API

With the New API, above code can be simplified as 

```java
return HttpResponses.BAD_REQUEST.error(UNKNOWN_AUTHENTICATOR_MSG_FORMAT, authenticatorName);
```

#### API lists

Below are the APIs that are provided by `HttpResponses`

```java
// returns a WebApplicationException that wraps a Response object
public RuntimeException exception(String message);

// returns a HTTP Response whose Content-Type is text/plain
public Response text(String message);

// returns an HTTP Response in error format. And the Content-Type is application/json
public Response error(String message);

// returns an HTTP Response in error format. And the Content-Type is application/json
public Response error(@Nullable Throwable t);

// returns an HTTP Response in JSON format with given object
public Response entity(@Nonnull Object entity)
```

### Incompatibility

Since this PR changes all current code to use the new API, there're some compatibility problems. Some HTTP endpoints now returns a JSON instead of returning a text-based response if the Response's status code is not 200. If clients have code to check the returned message for error processing, they need to change their code to adapt to the new returned format.


### What's not done in this PR

`Response.ok()` can be replaced by the new API `HttpResponses.OK.json()` so that we don't mix the old `Response` and new `HttpResponses` in the code. 

But it's not forbidden in this PR because the new one is just a replacement which won't introduce any compatibility problems. Since there're lots of reference of such API, and changes in this PR is already too large,  I think we can do that replacement in a following PR once this is merged.


<hr>

##### Key changed/added classes in this PR
 * `ResponseStatusException` is added to wrap any exceptions that want to return specific HTTP status code
 *  `ResponseStatusExceptionMapper` is provided to map any `ResponseStatusException` to a HTTP response which contains the error message in json format
 * `HttpResponses` defines a set of common used HTTP status codes and APIs to return a Response object.



This PR has:
- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:
```diff
@@ -177,25 +179,13 @@ public Response taskPost(final Task task, @Context final HttpServletRequest req)
+            return Response.ok(ImmutableMap.of("task", task.getId())).build();
+          }
+          catch (EntryExistsException e) {
+            throw new BadRequestException(StringUtils.format("Task[%s] already exists!", task.getId()));
```
Earlier, this was returning a response like `return Response.status(BAD_REQUEST).entity().build()`. Shouldn't we continue to return a `Response` here instead of throwing an exception?

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:
```diff
@@ -468,9 +458,7 @@ public Response getWorkerConfigHistory(
-        return Response.status(Response.Status.BAD_REQUEST)
-                       .entity(ImmutableMap.<String, Object>of("error", e.getMessage()))
-                       .build();
+        throw new BadRequestException(e.getMessage());
```
Same comment of change from returning a `Response` to throwing an exception.

FrankChen021 replied:
Same reason as above.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:
```diff
@@ -569,9 +557,7 @@ public Response getTasks(
-        return Response.status(Status.BAD_REQUEST)
-                       .entity(StringUtils.format("Invalid state : %s, valid values are: %s", state, API_TASK_STATES))
-                       .build();
+        throw new BadRequestException(StringUtils.format("Invalid state : %s, valid values are: %s", state, API_TASK_STATES));
```
Same comment of change from returning a `Response` to throwing an exception.

FrankChen021 replied:
I know it's an internal API. For the same reason above, I want to keep the code in the same way.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:
```diff
@@ -1637,30 +1631,23 @@ public Response setEndOffsets(
         } else if (latestSequence.isCheckpointed()) {
-          return Response.status(Response.Status.BAD_REQUEST)
-                         .entity(StringUtils.format(
+          return ResponseStatusExceptionMapper.toResponse(Response.Status.BAD_REQUEST,
```
_Nit: style choice_
Maybe break the line after `.toResponse(`, would look more readable if the arguments start from a new line.

kfaraz commented on indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java:
```diff
@@ -1637,30 +1631,23 @@ public Response setEndOffsets(
-                               )
-                           )
-                           .build();
+            return ResponseStatusExceptionMapper.toResponse(Response.Status.BAD_REQUEST,
```
_Nit: same format suggestion as above_

kfaraz commented:
Thanks for the changes, @FrankChen021!
This makes the error responses much cleaner!

I had a couple of minor suggestions:
- In `ResponseStatusExceptionMapper`, consider changing the signature of method `.toResponse()` to include `Object... args` so that the `String.format` can be done there instead of every caller having to do it.
- Consider adding a syntax sugar for bad request and not found (as these are most commonly used) over `.toResponse()` i.e. something like `.toBadRequestResponse()` so that the caller can skip passing the `Response.Status.BAD_REQUEST` in the common cases.

FrankChen021 commented:
@kfaraz My opinion is that we should use exception-thrown style instead of returning statement. The reason I keep a `toResponse` function to return a `Response` object is that there're lots of code using returning statement, and to change them all will introduce a burden of test.

jihoonson commented:
Hey @FrankChen021, can you add some tests to verify the new response?

FrankChen021 commented:
@jihoonson This patch updates many returning statements, it's hard to add cases for all these changes.

For the tests:

I added a UT test case to check the returned `Response` object, including the Header, Object and Status Code; can check this file: indexing-service/src/test/java/org/apache/druid/indexing/overlord/http/OverlordResourceTest.java

I also updated current integration test to check if the `ResponseStatusExceptionMapper` successfully coverts exception into `Response` object and returned Header, Content and Status are correct. See:

integration-tests/src/test/java/org/apache/druid/tests/query/ITOverlordResourceTest.java

FrankChen021 commented:
@kfaraz I added an override version to accept `Object... args` parameters to simplify. And also I added some comments to these functions to give suggestions that when 400 or 404 is expected to return, exception throwing style is recommended.

FrankChen021 commented:
CI failed because of insufficient line coverage.

jihoonson reviewed (commented):
@FrankChen021 thank you for fixing this bug. The new framework you added to consolidate the way to handle responses looks good to me. 

> @kfaraz My opinion is that we should use exception-thrown style instead of returning statement. The reason I keep a `toResponse` function to return a `Response` object is that there're lots of code using returning statement, and to change them all will introduce a burden of test.

Why do you think we _should_ use exceptions instead of using whatever the easier between them in each case? I would want to just use whatever the easier one.

> @jihoonson This patch updates many returning statements, it's hard to add cases for all these changes.

Thank you for adding new tests. But I don't think we can merge this PR without a green mark from Travis. I understand there are many changes to test. I also understand that you are blocked by lack of tests because we haven't added any test for them yet which would have been better if the original author who wrote this code added required tests in the first place. However, please understand that it is also very hard to enforce 100% test coverage for every contributor. It will slow down our development speed significantly. What we can do is making things better together over time. Lack of test coverage from less-relevant changes happens to everyone including myself. We should not ignore them but fix them. 

Looking at the travis failures, it doesn't seem actually that bad.

```
Total diff coverage:

 - lines: 46% (15/32)
 - branches: 100% (2/2)
 - functions: 100% (3/3)
 
ERROR: Insufficient line coverage of 46% (15/32). Required 50%.
```

```
Total diff coverage:

 - lines: 47% (21/44)
 - branches: 33% (2/6)
 - functions: 43% (27/62)

ERROR: Insufficient line coverage of 47% (21/44). Required 50%.
ERROR: Insufficient branch coverage of 33% (2/6). Required 50%.
```

You can just add a couple of more tests to make it pass, even though I would encourage you to add as many tests as possible.

FrankChen021 commented on indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexSupervisorTask.java:
```diff
@@ -1297,7 +1298,7 @@ public Response getPhaseName(@Context final HttpServletRequest req)
       }
     } else {
-      return Response.status(Status.BAD_REQUEST).entity("task is running in the sequential mode").build();
+      return ResponseStatusExceptionMapper.toResponse(Response.Status.BAD_REQUEST, "task is running in the sequential mode");
```
I cleaned up all the BAD_REQUEST response code to use the same response format.

FrankChen021 replied:
I think it's safe to do this because we don't know in the future if the string content will be changed. If it's changed, there's a risk to experience the same problem we're facing now.

Actually, the original API reported by the author is also a user facing API, changing the response seems incompatibility. But I think it's OK because these changes happen in error cases which are rare.

jihoonson replied:
Perhaps I should have been clearer. All I'm worrying about is the error message. It should not change unless there is a good reason especially for user-facing APIs.

jihoonson replied:
Let me make my stance clearer here. I think `ResponseStatusExceptionMapper` is a good idea to prevent future mistakes of missing content type. It's worth modifying user-facing API responses. To make sure that we always use the nice helper method instead of making a `Response` on our own, we should probably register `Response.status()` or `Response.serverError()` as forbidden APIs. If we are going down that way, we should change all user-facing APIs in the same release so that we won't introduce similar incompatible changes over multiple releases. Perhaps you or someone else is planning to do this in multiple PRs, but because things don't always go as planned, I think it should be better to do in one PR. But this is going to be a huge refactoring, so I would suggest to do this separately from the original security issue.

FrankChen021 replied:
I intendend to forbidden the call of `Response.status` or `Response.serverError()`. 

As you said, there are too many references we need to modify, so I didn't do all the things in this patch, but just modified all BAD_REQUEST, which was the same error as the original report, and 3 or 4 places of NOT_FOUND error.

FrankChen021 commented on indexing-service/src/main/java/org/apache/druid/indexing/overlord/http/OverlordResource.java:
```diff
@@ -587,11 +574,7 @@ public Response getTasks(
-                    .entity(StringUtils.format("Access-Check-Result: %s", authResult.toString()))
-                    .build()
-        );
+        throw new ForbiddenException(StringUtils.format(authResult.toString()));
```
No, I changed it. There's still a risk here if we don't set the Content-Type.

jihoonson replied:
Why did you change the error message then?

FrankChen021 replied:
Ha, I got what you mean, this is changed accidentally.

FrankChen021 commented on server/src/main/java/org/apache/druid/server/initialization/jetty/ResponseStatusExceptionMapper.java:
```diff
@@ -0,0 +1,71 @@
+{
+  /**
+   * For Response.Status.BAD_REQUEST or Response.Status.NOT_FOUND,
+   * it's suggested to throw exception {@link BadRequestException} or {@link NotFoundException} respectively
```
My first thought is that we should apply the same error-handling style for all code, and the `ResponseStatusExceptionManager` maps the exception to a response so that user don't need to care about the `Response` object.

Considering lots of code returns the `Response` object directly, and it seems no gain to change all these code to exception-throwing style, and also seems no one will do this refactoring at any time.

So, I removed this in the newer commit.

FrankChen021 commented:
@jihoonson Could you review this PR when you're at convenience? I saw there were conflicts with the latest changes on master, and worried about more conflicts.

jihoonson commented:
@FrankChen021 my apologies, I forgot to take another look. Please feel free to ping me if I forget again and don't finish my review in time.

I took another look. Thank you for adding more tests to satisfy the coverage bot :slightly_smiling_face: The general approach you took in this PR looks good to me. It's much better-structured compared to what we had previously. However, I still have a concern about the user-facing API changes, which I mentioned in https://github.com/apache/druid/pull/12026#discussion_r770186406. I agree that your approach is better than what we have now because it makes the error response format consistent. I think we should apply the same approach to every user-facing API. And I think this should happen at once because it will cause incompatibility in the error response format for some APIs. Suppose you have a client that does something when an API returns an error. For example, it could retry the failed call. If the error response format changes after this PR, you will have to modify your client code to handle the new response format. Even though the new response format is better than the previous, it will be annoying our users if we introduce such an incompatible change over multiple releases. Since we will probably start preparing a new release sooner or later, we may not be able to have enough time to apply the same technique you used in this PR to every user-facing API. So, how about splitting your PR into two, one that fixes the vulnerability only and another that fixes the error response format?

FrankChen021 commented:
@jihoonson A good suggestion. Let me create a new PR to fix the vulnerability only and leave this PR to unify the response model.

jihoonson commented:
@FrankChen021 thanks for understanding. I will review #12295 soon.

FrankChen021 commented:
Description has been updated to reflect the latest change. And the old API will be added to the forbidden API list in later commits.

FrankChen021 commented:
@kfaraz @jihoonson Could you please take a look at the changes in this PR when you're at convenience?

kfaraz commented:
Sure, @FrankChen021 . Sorry for the delay on this. Will take a look this week.
